proposals/text/0000-evaluation-haystack-2.md

```diff
@@ -0,0 +1,227 @@
+
+```python
+def eval(runnable: Union[Pipeline, Component], inputs: List[Dict[str, Any]], expected_outputs: List[Dict[str, Any]]) -> EvaluationResult:
+    outputs = []
+    for input_ in inputs:
+      output = runnable.run(input_)
+      outputs.append(output)
+    return EvaluationResult(runnable, inputs, outputs, expected_outputs)
+```
+
+This is obviously an overtly simplistic example but the core concept remains.
+`inputs` must be a list of data that will be passed to either the `Pipeline` or the `Component`.
+`expected_outputs` could be a list with the same length of `inputs` or an empty list for blind evaluation.
+
+Blind in this context means running an evaluation without providing a list of expected output. This could be done for several reasons, like if we don't know what to expect as output, or to compare output of different components.
+
+`EvaluationResult` could either be a `Dict` or its own class, this is open to discussion. Either way it must be easy to save to disk. When saving the results to disk we can also include the `Pipeline` or `Component` in a serialized form.
+
+When evaluating a `Pipeline` we could also override its private `_run_component` function to evaluate every node it will run. This will 100% work for our implementation of `Pipeline`. If a user tries to evaluate a `Pipeline` that reimplements its own `run` method it might not be able to evaluate each `Component`. I believe this a worthy risky tradeoff.
+
+Overriding `_run_component` would also give us the chance to simulate optimal component outputs. `eval` could also accept an optional `simulated_output` dictionary containing the outputs of one or more `Component` that are in the `Pipeline`. It would look similar to this:
+
+```python
+simulated_output = {
+  "component_name": {"answer": "120"},
+  "another_component_name": {"metadata": {"id": 1}}
+}
+```
+
+Another alternative would be to use the `expected_output` also to simulate outputs of intermediate components.
+
+#### Tracking progress
+
+To track progress we can also go on another direction. We could return partials results while iterating our inputs:
+
+```
+def eval(runnable, inputs, expected_outputs):
+    result = EvaluationResult(runnable, inputs, {}, expected_outputs)
+    for input_ in inputs:
+      output = runnable.run(input_)
+      result.append_output(output)
+      yield result
+```
+
+Or return some progress percentage and only in the end the final result.
+
+```
+def eval(runnable, inputs, expected_outputs):
+    outputs = []
+    total = len(inputs)
+    for i, input_ in enumerate(inputs):
+      output = runnable.run(input_)
+      outputs.append(output)
+      yield 100 * (i / total), None
+
+    yield 100, EvaluationResult(runnable, inputs, outputs, expected_outputs)
+```
+
+### `EvaluationResult`
+
+`EvaluationResult` won't contain partial metrics anymore as it does in v1.x but it will keep track of all the information used by `eval()`.
```

Comment:
Actually I think there's one major source of information that this design doesn't take into account, which is the Document Store's content. It is the main input of a Retriever for example, and yet EvaluationResult won't store any information about it, making replication of the same experiment impossible with this object and the Pipeline alone.

I don't think we should address the problem in this setting, but let's add a note stating we're aware of the limitation.

Reply:
I think addressing this problem would be close to impossible. The only information we should provide about a DocumentStore is number of documents stored in it. Providing the content would mean copying the whole DocumentStore data and it can snowball easily. It could work with 10 Documents, what about 100? Or 100.000?

Also in the proposal when talking about reproducibility I'm not talking about the evaluation experiment but the metrics. The goal is that it should be easy to recalculate metrics without making it necessary to rerun the `Pipeline`.